### PR TITLE
Fix SELU gradient precision for negative inputs

### LINKED ISSUE #124834: tf.nn.selu returns wrong gradient for large negative float64 input
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

 Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.selu` returns an incorrect gradient for a large negative float64 input.

For SELU, the negative branch is a scaled ELU branch. At `x = -700.0`, the analytic high-precision reference for the derivative is approximately `1.7334290832552396e-304`, which is a finite nonzero float64 value.

However, TensorFlow autodiff returns `0.0` for this gradient. PyTorch computes the expected finite value for the same derivative, so this appears to be a backward-pass issue rather than a float64 limitation.

Expected behavior?

The gradient of `tf.nn.selu` at `x = -700.0` should be close to the analytic reference value `1.7334290832552396e-304`, not `0.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

x = -700.0
truth = 1.7334290832552396e-304
rtol = 1e-6

v = tf.Variable(x, dtype=tf.float64)

with tf.GradientTape() as tape:
    y = tf.nn.selu(v)

g = tape.gradient(y, v)

actual = float(g)
rel_err = abs(actual - truth) / abs(truth)

print(f"x: {x!r}")
print(f"tensorflow gradient: {actual!r}")
print(f"expected gradient: {truth!r}")
print(f"relative error: {rel_err:.3e}")

if rel_err > rtol:
    print("BUG REPRODUCED: tf.nn.selu gradient is wrong for large negative float64 input")
else:
    print("not reproduced")
```

### Relevant log output

```shell
x: -700.0
tensorflow gradient: 0.0
expected gradient: 1.7334290832552396e-304
relative error: 1.000e+00
BUG REPRODUCED: tf.nn.selu gradient is wrong for large negative float64 input
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
Fixes #124834.

## What changed

- Compute the SELU derivative from the original input instead of reconstructing it from the rounded activation output.
- Preserve the existing positive, signed-zero, infinity, and NaN behavior while using the stable `scale_alpha * exp(x)` negative branch.
- Update generated GradientTape exclusion metadata so eager execution retains the input and releases the now-unused output.
- Add float32 and float64 regression coverage in both graph and eager modes.

## Validation

- Reproduced the bug on TensorFlow 2.20.0: both float32 at `-20` and float64 at `-700` return a zero gradient.
- Verified the patched registered gradient returns `3.6237128853144895e-09` and `1.7334290832552396e-304`, respectively, in graph mode and direct eager execution.
- Verified vector inputs, infinities, signed zero, positive inputs, and NaN semantics against the analytic SELU derivative.
- Verified TensorFlow's gradient metadata analysis reports live inputs `{0}` and no live outputs for `_SeluGrad`.
- `python3 -m py_compile tensorflow/python/ops/nn_grad.py tensorflow/python/ops/nn_grad_test.py`
- TensorFlow PyLint configuration: 10.00/10 (excluding two unrelated existing inference false positives and an obsolete-option warning in the repository config under the current PyLint release).
- `git diff --check`

### CODE REVIEW

File: tensorflow/python/ops/nn_grad_test.py [hunk 1]
```diff
@@ -218,6 +218,26 @@ def testEluGradGradWRTinputs(self):
...
+        with backprop.GradientTape() as tape:
+          tape.watch(inputs)
+          selu = gen_nn_ops.selu(inputs)
+
+        selu_grad = tape.gradient(selu, inputs)
+        np_dtype = dtype.as_numpy_dtype
+        expected = np_dtype(scale_alpha) * np.exp(np_dtype(value))
+        self.assertAllClose(
+            expected, self.evaluate(selu_grad), rtol=rtol, atol=0)
+
```
[the remark this answers is not in the document]
Comment (GodlyDonuts) [reply tensorflow/python/ops/nn_grad_test.py:259]: Addressed in 4d7fdb927a1. Added a graph/eager vector regression test for both float32 and float64 covering finite negative and positive inputs, signed zero, positive and negative infinity, and NaN semantics.

**Comment by GodlyDonuts** [comment]
@dmiltr3, would you be able to review this change when you have a chance? Thank you.

**Review by dmiltr3 — APPROVED** [review]
LGTM, thank you for the fix and thorough test coverage!

**Comment by GodlyDonuts** [comment]
The failing ARM64 and CUDA 13 checks appear infrastructural/unrelated, and the Windows failure appears to be the existing symbol-export issue addressed by #125122. @nithyak0204 Can you rerun the CI? Thanks

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
### Copybara & Presubmit Status

The `feedback/copybara` check failed because the imported presubmit pipeline requires a refresh against the latest `master` baseline.

#### Action required
Please rebase your branch on the latest `master` to trigger a fresh import and presubmit cycle:

```bash
git fetch upstream master
git rebase upstream/master
git push --force-with-lease
```

**Comment by GodlyDonuts** [comment]
Done! Sorry I took so long, I was busy with classes.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
### Summary of Failure
Internal strict dependency analysis checks failed when building `//tensorflow/python/ops:nn_grad`:

```
Strict deps violations: //tensorflow/python/ops:nn_grad
  In tensorflow/python/ops/nn_grad.py, no direct deps found for imports:
    from tensorflow.python.framework import constant_op
```

### Root Cause
`tensorflow/python/ops/nn_grad.py` imports `constant_op` to define the SELU scale constants, but `//tensorflow/python/ops:nn_grad` in `tensorflow/python/ops/BUILD` does not list `"//tensorflow/python/framework:constant_op"` in its `deps`.

### Required Fix
In `tensorflow/python/ops/BUILD`, add `"//tensorflow/python/framework:constant_op"` to the `deps` list of `py_library(name = "nn_grad")`:

```diff
--- a/tensorflow/python/ops/BUILD
+++ b/tensorflow/python/ops/BUILD
@@ -2231,6 +2231,7 @@ py_library(
         ":array_ops_stack",
         ":math_ops",
         ":nn_ops_gen",
+        "//tensorflow/python/framework:constant_op",
         "//tensorflow/python/framework:dtypes",
         "//tensorflow/python/framework:ops",
     ],
```

**Review by dmiltr3 — APPROVED** [review]
LGTM, thank you for adding the missing dependency and updating the tests!
